### DISCUSSION TITLE
How to Migrate from Supabase Auth Helpers to SSR package

### DISCUSSION THREAD

**Discussion by ssudaraka** [supabase/supabase#27849] [Troubleshooting]
_This is a copy of a troubleshooting article on Supabase's docs site. It may be missing some details from the original. View the [original article](https://supabase.com/docs/guides/troubleshooting/how-to-migrate-from-supabase-auth-helpers-to-ssr-package-5NRunM)._

The `auth-helpers` packages are deprecated and replaced with the `@supabase/ssr` package. We recommend migrating to the `@supabase/ssr` package as future bug fixes and feature releases are focused on the `@supabase/ssr` package.

Here are the steps for you to migrate your application from the `auth-helpers` package to `@supabase/ssr` package.

Depending on your implementation, you may ignore some parts of this documentation and use your own implementation (i.e. using API routes vs. Server Actions). What's important is you replace the clients provided by `auth-helpers` with the utility functions created using clients provided by `@supabase/ssr`.

## 1. Uninstall Supabase Auth helpers and install the Supabase SSR package

It's important that you don't use both `auth-helpers-nextjs` and `@supabase/ssr` packages in the same application to avoid running into authentication issues.

    npm uninstall @supabase/auth-helpers-nextjs @supabase/supabase-js
    npm install @supabase/ssr @supabase/supabase-js

## 2. Create the library functions to create Supabase clients

    // lib/supabase/client.ts

    import { createBrowserClient } from '@supabase/ssr';

    export function createClient() {
      return createBrowserClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY!
      );
    }

    // lib/supabase/server.ts
    import { createServerClient } from '@supabase/ssr'
    import { cookies } from 'next/headers'

    export async function createClient() {
      const cookieStore = await cookies()

      return createServerClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY!,
        {
          cookies: {
            getAll() {
              return cookieStore.getAll()
            },
            setAll(cookiesToSet, _headers) {
              try {
                cookiesToSet.forEach(({ name, value, options }) => cookieStore.set(name, value, options))
              } catch {
                // The `setAll` method was called from a Server Component.
                // This can be ignored if you have middleware refreshing
                // user sessions.
              }
            },
          },
        }
      )
    }

    // lib/supabase/proxy.ts
    import { createServerClient } from '@supabase/ssr'
    import { NextResponse, type NextRequest } from 'next/server'

    export async function updateSession(request: NextRequest) {
      let supabaseResponse = NextResponse.next({
        request,
      })

      // With Fluid compute, don't put this client in a global environment
      // variable. Always create a new one on each request.
      const supabase = createServerClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY!,
        {
          cookies: {
            getAll() {
              return request.cookies.getAll()
            },
            setAll(cookiesToSet, headers) {
              cookiesToSet.forEach(({ name, value }) => request.cookies.set(name, value))
              supabaseResponse = NextResponse.next({
                request,
              })
              cookiesToSet.forEach(({ name, value, options }) => supabaseResponse.cookies.set(name, value, options))
              Object.entries(headers).forEach(([key, value]) =>
                supabaseResponse.headers.set(key, value)
              )
            },
          },
        }
      )

      // Do not run code between createServerClient and
      // supabase.auth.getClaims(). A basic mistake could make it very hard to debug
      // issues with users being randomly logged out.

      // IMPORTANT: If you remove getClaims() and you use server-side rendering
      // with the Supabase client, your users may be randomly logged out.
      const { data } = await supabase.auth.getClaims()

      const user = data?.claims

      if (
        !user &&
        !request.nextUrl.pathname.startsWith('/login') &&
        !request.nextUrl.pathname.startsWith('/auth')
      ) {
        // no user, potentially respond by redirecting the user to the login page
        const url = request.nextUrl.clone()
        url.pathname = '/login'
        return NextResponse.redirect(url)
      }

      // IMPORTANT: You *must* return the supabaseResponse object as it is. If you're
      // creating a new response object with NextResponse.next() make sure to:
      // 1. Pass the request in it, like so:
      //    const myNewResponse = NextResponse.next({ request })
      // 2. Copy over the cookies, like so:
      //    myNewResponse.cookies.setAll(supabaseResponse.cookies.getAll())
      // 3. Change the myNewResponse object to fit your needs, but avoid changing
      //    the cookies!
      // 4. Finally:
      //    return myNewResponse
      // If this is not done, you may be causing the browser and server to go out
      // of sync and terminate the user's session prematurely!

      return supabaseResponse
    }

## 3. Replace your proxy.ts file

    // proxy.ts

    import { type NextRequest } from "next/server"
    import { updateSession } from "@/lib/supabase/proxy"

    export async function proxy(request: NextRequest) {
      return await updateSession(request)
    }

    export const config = {
      matcher: [
        /*
         * Match all request paths except for the ones starting with:
         * - _next/static (static files)
         * - _next/image (image optimization files)
         * - favicon.ico (favicon file)
         * Feel free to modify this pattern to include more paths.
         */
        "/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)",
      ],
    }

## 4. Create your server actions to handle login and sign up

    // app/login/actions.ts

    'use server';

    import { revalidatePath } from 'next/cache';
    import { redirect } from 'next/navigation';

    import { createClient } from '@/lib/supabase/server';

    export async function login(formData: FormData) {
      const supabase = createClient();

      // type-casting here for convenience
      // in practice, you should validate your inputs
      const data = {
        email: formData.get('email') as string,
        password: formData.get('password') as string,
      };

      const { error } = await supabase.auth.signInWithPassword(data)

      if (error) {
        redirect('/error');
      }

      revalidatePath('/', 'layout');
      redirect('/');
    }

    export async function signup(formData: FormData) {
      const supabase = createClient();

      // type-casting here for convenience
      // in practice, you should validate your inputs
      const data = {
        email: formData.get('email') as string,
        password: formData.get('password') as string,
      };

      const { error } = await supabase.auth.signUp(data);

      if (error) {
        redirect('/error');
      }

      revalidatePath('/', 'layout');
      redirect('/');
    }

## 5. Use the server actions in your login page UI

    // app/login/page.tsx

    import { login, signup } from './actions';

    export default function LoginPage() {
      return (
        <form>
          <label htmlFor="email">Email:</label>
          <input id="email" name="email" type="email" required />
          <label htmlFor="password">Password:</label>
          <input id="password" name="password" type="password" required />
          <button formAction={login}>Log in</button>
          <button formAction={signup}>Sign up</button>
        </form>
      );
    }

## 6. Client components

    'use client';

    // replace this line
    import { createClientComponentClient } from '@supabase/auth-helpers-nextjs';

    // with
    import { createClient } from '@/lib/supabase/client';

    export default async function Page() {
     // replace this line
     const supabase = createClientComponentClient<Database>();

     // with
     const supabase = createClient();

     return...
    }

## 7. Server components

    // replace
    import { cookies } from 'next/headers';
    import { createServerComponentClient } from '@supabase/auth-helpers-nextjs';

    // with
    import { createClient } from '@/lib/supabase/server';

    export default async function Page() {
     // replace
     const cookieStore = cookies();
     const supabase = createServerComponentClient<Database>({
      cookies: () => cookieStore
     });

     // with
     const supabase = createClient();

     return...
    }

## 8. Route handlers

    // replace
    import { createRouteHandlerClient } from '@supabase/auth-helpers-nextjs';
    import { cookies } from 'next/headers';

    // with
    import { createClient } from '@/lib/supabase/server';

    export async function POST(request: Request) {
     // replace
     const supabase = createRouteHandlerClient<Database>({
        cookies: () => cookieStore,
      });

      // with
      const supabase = createClient();

      return...
    }

Likewise, you can replace the clients created with `@supabase/auth-helpers-nextjs` with utility functions you created with `@supabase/ssr`.

`createMiddlewareClient` → `createServerClient`
`createClientComponentClient` → `createBrowserClient`
`createServerComponentClient` → `createServerClient`
`createRouteHandlerClient` → `createServerClient`

You can find more clear and concise examples of creating clients [in our SSR documentation](https://supabase.com/docs/guides/auth/server-side/creating-a-client?queryGroups=framework\&framework=nextjs\&queryGroups=environment\&environment=route-handler#creating-a-client).

If you have any feedback about this guide, provide them as a comment below. If you find any issues or have feedback for the `@supabase/ssr` client, post them as an issue in `@supabase/ssr` repo.

As always, our GitHub community and Discord channel are open for technical discussions and resolving your issues.

**Comment by renchris** [comment]
Supabase SSR in NextJS Middleware is not compatible with NextJS 15. Cookies are not being saved because Supabase is not correctly handling cookies that are differently handled in NextJS 15.

We had to downgrade React dependencies to latest 18 version `18.3.1` and NextJS dependencies to latest 14 version `14.2.20` for this to work.

Would we be able to get the examples and documentation updated to work with NextJS 15 + React 19?

**Reply by kahouajiomar** [reply]
any updates ? :(

**Reply by enlguy** [reply]
Having issues with this as well, amazing they still haven't added support for NextJS 15 in October of 2025.  Supabase is becoming a bastard stepchild in my book, it is always overly complicated to work with.  Commented on another major unresolved bug yesterday.

**Reply by RobSchilderr** [reply]
Isn't this because you're overwriting something? @enlguy 

We got Supabase working with Next.js 15 and app router. Check the Supabase jwt-signing-keys example
